cranelift/interpreter/src/interpreter.rs

```diff
@@ -440,4 +499,209 @@ mod tests {
+        function %callee(i64, i64) -> i64 {
+            ss0 = explicit_slot 8
+            ss1 = explicit_slot 8
+
+        block0(v0: i64, v1: i64):
+            stack_store.i64 v0, ss0
+            stack_store.i64 v1, ss1
+            v2 = stack_load.i64 ss0
+            v3 = stack_load.i64 ss1
+            v4 = iadd.i64 v2, v3
+            return v4
+        }
+
+        function %caller(i64, i64, i64, i64) -> i64 {
+            fn0 = %callee(i64, i64) -> i64
+            ss0 = explicit_slot 8
+            ss1 = explicit_slot 8
+
+        block0(v0: i64, v1: i64, v2: i64, v3: i64):
+            stack_store.i64 v0, ss0
+            stack_store.i64 v1, ss1
+
+            v4 = call fn0(v2, v3)
+
+            v5 = stack_load.i64 ss0
+            v6 = stack_load.i64 ss1
+
+            v7 = iadd.i64 v4, v5
+            v8 = iadd.i64 v7, v6
+
+            return v8
+        }";
+
+        let mut env = FunctionStore::default();
+        let funcs = parse_functions(code).unwrap().to_vec();
+        funcs.iter().for_each(|f| env.add(f.name.to_string(), f));
+
+        let state = InterpreterState::default().with_function_store(env);
+        let result = Interpreter::new(state)
+            .call_by_name(
+                "%caller",
+                &[
+                    DataValue::I64(3),
+                    DataValue::I64(5),
+                    DataValue::I64(7),
+                    DataValue::I64(11),
+                ],
+            )
+            .unwrap()
+            .unwrap_return();
+
+        assert_eq!(result, vec![DataValue::I64(26)])
+    }
+
+    #[test]
+    fn out_of_slot_write_traps() {
+        let code = "
+        function %stack_write() {
+            ss0 = explicit_slot 8
+
+        block0:
+            v0 = iconst.i64 10
+            stack_store.i64 v0, ss0+8
+            return
+        }";
+
+        let func = parse_functions(code).unwrap().into_iter().next().unwrap();
+        let mut env = FunctionStore::default();
+        env.add(func.name.to_string(), &func);
+        let state = InterpreterState::default().with_function_store(env);
+        let trap = Interpreter::new(state)
+            .call_by_name("%stack_write", &[])
+            .unwrap()
+            .unwrap_trap();
+
+        assert_eq!(trap, CraneliftTrap::User(TrapCode::InvalidAddress));
+    }
+
+    #[test]
+    fn partial_out_of_slot_write_traps() {
+        let code = "
+        function %stack_write() {
+            ss0 = explicit_slot 8
+
+        block0:
+            v0 = iconst.i64 10
+            stack_store.i64 v0, ss0+4
+            return
+        }";
+
+        let func = parse_functions(code).unwrap().into_iter().next().unwrap();
+        let mut env = FunctionStore::default();
+        env.add(func.name.to_string(), &func);
+        let state = InterpreterState::default().with_function_store(env);
+        let trap = Interpreter::new(state)
+            .call_by_name("%stack_write", &[])
+            .unwrap()
+            .unwrap_trap();
```

Comment:
I should really create a helper functions for this common "just run the interpreter on this CLIF" case. Don't worry about it for this PR but eventually I'll do something like that.

Reply:
Yeah, sounds like a good idea. I can probably fix that when I implement heap access, ill probably have to add a number of tests similar to these ones.